### ISSUE TITLE
Deployment container or image name change ignored when server-side apply is enabled.

### ISSUE THREAD

**Issue by MitchellGerdisch** [pulumi/pulumi-kubernetes#2322]
### What happened?

After deploying a k8s deployment that uses server-side apply, and then changing the container or image name, Pulumi does not detect a change when running `pulumi up`.

### Expected Behavior

Changes to the pulumi code should be detected and updates made accordingly.

### Steps to reproduce

Using the code in this file: [__main__.py.txt](https://github.com/pulumi/pulumi-kubernetes/files/10767268/__main__.py.txt)

1. Deploy a stack using the above code
2. modify `container_name` and/or `image_name` (around line 80)
3. Run `pulumi preview` and/or `pulumi up` and note that no change is detected when in fact the code has changed.

### Output of `pulumi about`

CLI          
Version      3.55.0
Go Version   go1.20
Go Compiler  gc

Plugins
NAME        VERSION
gcp         6.49.0
kubernetes  3.24.0
python      unknown
random      4.11.2

Host     
OS       darwin
Version  12.5.1
Arch     x86_64

Found no pending operations associated with dev

Backend        
Name           pulumi.com

Dependencies:
NAME               VERSION
pip                23.0.0
pulumi-gcp         6.49.0
pulumi-kubernetes  3.24.0
pulumi-random      4.11.2
setuptools         67.3.2
wheel              0.38.4

### Additional context

Things work correctly (i.e. updates are detected and executed) when server-side apply is NOT enabled.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by squaremo** [comment]
I was able to see this by using `pulumi new kubernetes-typescript` and adapting a little:

```
import * as k8s from "@pulumi/kubernetes";
import * as kx from "@pulumi/kubernetesx";

const provider = new k8s.Provider("local", { enableServerSideApply: true });

const appLabels = { app: "nginx" };
const deployment = new k8s.apps.v1.Deployment("nginx", {
    spec: {
        selector: { matchLabels: appLabels },
        replicas: 1,
        template: {
            metadata: { labels: appLabels },
            spec: { containers: [{ name: "nginx1", image: "nginx" }] }
        }
    }
}, { provider });

export const name = deployment.metadata.name;
```

If I `pulumi up`, then change the name of the container, and `pulumi up` again, it says nothing has changed and takes no actions.

**Comment by lblackstone** [comment]
This works in the v4 release when you run a refresh in either CSA or SSA mode (fixed in #2445).
